base/linalg/lq.jl

```diff
@@ -88,22 +88,25 @@ end
 
 size(A::LQPackedQ) = size(A.factors)
 
-full(A::LQ) = A[:L]*A[:Q]
-#=
-We construct the full eye here, even though it seems ineffecient, because
-every element in the output matrix is a function of all the elements of
-the input matrix. The eye is modified by the elementary reflectors held
-in A, so this is not just an indexing operation. Note that in general
-explicitly constructing Q, rather than using the ldiv or mult methods,
-may be a wasteful allocation.
-=#
-function full{T}(A::LQPackedQ{T}; thin::Bool=true)
+convert(::Type{Matrix}, A::LQ) = convert(Array, A[:L]*A[:Q])
+convert(::Type{Array}, A::LQ) = convert(Matrix, A)
+full(A::LQ) = convert(Array, A)
+
+function convert{T}(::Type{Matrix}, A::LQPackedQ{T}; thin::Bool = true)
```

Comment:
Sorry if I misunderstood. To make certain we are on the same page:

`qr` by default returns a thin QR factorization as a pair of matrices, and provides the option to simultaneously expand `Q` without expanding `R`. IIUC, `@andreasnoack` would like to retain this behavior. This series of PRs does not impact `qr`'s behavior.

`qrfact` by default returns a thin factorization as `F <: Factorization`. When the user extracts `Q` from `F`, the user receives a `QRPackedQ` or `QCCompactWYQ` type that efficiently represents the thin `Q`, and with which the user can perform e.g. multiplication and have everything work without thought. (I.e., depending on the operation, the thin `Q` in some cases behaves like a (strict) thin `Q` and in others like a square `Q`). IIUC, `@andreasnoack` would like to retain this behavior. This series of PRs impacts neither `qrfact`  nor `QRPackedQ`/`QRCompactWYQ`'s behavior.

`full` on `QRPackedQ`/`QRCompactWYQ` by default returns the thin `Q` that those types represent as an `Array`, and provides the option to simultaneously expand `Q`. This series of PRs does not impact `full`'s behavior preceding its deprecation, and the replacement `convert` provides `full`'s existing default behavior. This is the hazy point.

On the one hand, having `convert(Array, X::Union{QRPackedQ,QRCompactWYQ})` return an explicit representation of the thin `Q` that `X` represents seems natural given the name `convert`; anything more is not merely `convert`ing `X` to an `Array`.

On the other hand, as `@andreasnoack` points out, having `full` return an expanded `Q`, in contrast to its present default behavior, might be more consistent with the behavior of `QRPackedQ`/`QRCompactWYQ` under multiplication, and hence might be a better interpretation of `full` in this context.

This haziness strikes me as another consequence of `full`'s ambiguous meaning. In this context, `full` could mean either strict `convert` or "return an `Array` that behaves like `QRPackedQ`/`QRCompactWYQ` under certain operations", muddling two orthogonal concepts/operations.

I advocate we do away with this ambiguity, decoupling the orthogonal concepts and naming them appropriately: have `convert` strictly `convert`, and provide another operation, say `completeQ`, `basisQ`, or something such, that generates an orthogonal/unitary matrix from a rectangular matrix containing a set of orthogonal vectors (also applicable to `LQPackedQ` and perhaps generally useful in its own right). Doing so seems like it would yield a better point in simplicity-clarity-generality space than not separating these concepts.

Thoughts? Best!